apps/src/lib/tools/jsinterpreter/patchInterpreter.js

```diff
@@ -0,0 +1,99 @@
+const Interpreter = require('@code-dot-org/js-interpreter');
+
+module.exports = function patchInterpreter() {
+
+  const base = {
+    hasProperty: Interpreter.prototype.hasProperty,
+    getProperty: Interpreter.prototype.getProperty,
+    setProperty: Interpreter.prototype.setProperty,
+  };
+
+  // These methods need to be patched in order to support custom marshaling.
+
+  // These changes revert a 10% speedup commit that bypassed hasProperty,
+  // getProperty, and setProperty:
+  // https://github.com/NeilFraser/JS-Interpreter/commit/c6f25b4a30046a858e5e90a92a8c0d24a93c0231
+
+  /**
+   * Retrieves a value from the scope chain.
+   * @param {!Object} name Name of variable.
+   * @return {!Object} The value.
+   */
+  Interpreter.prototype.getValueFromScope = function (name) {
+    var scope = this.getScope();
+    var nameStr = name.toString();
+    while (scope) {
+      if (this.hasProperty(scope, nameStr)) {
+        return this.getProperty(scope, nameStr);
+      }
+      scope = scope.parentScope;
+    }
+    this.throwException('Unknown identifier: ' + nameStr);
+    return this.UNDEFINED;
```

Comment:
It throws an exception _sometimes_. The case where it doesn't is definitely not covered here!

Though frankly, this code is only really covered by integration tests, so I really need to go back and add unit tests.